File: analysis/import/CSVImporter/src/test/java/de/maibornwolff/codecharta/importer/csv/CSVHeaderTest.java
package de.maibornwolff.codecharta.importer.csv;

import com.google.common.collect.ImmutableSet;
import org.junit.Test;

import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.assertThat;

public class CSVHeaderTest {
    @Test(expected = IllegalArgumentException.class)
    public void emptyHeader_should_throw_exception() {
        new CSVHeader(new String[]{});
    }

    @Test
    public void columnNumbers_should_return_all_columnNumbers() {
        CSVHeader header = new CSVHeader(new String[]{"first", "second"});
        assertThat(header.getColumnNumbers(), is(ImmutableSet.of(0, 1)));
    }

    @Test
    public void empty_header_columns_should_be_ignored() {
        CSVHeader header = new CSVHeader(new String[]{"first", "second", ""});
        assertThat(header.getColumnNumbers(), is(ImmutableSet.of(0, 1)));
    }

    @Test
    public void null_in_header_columns_should_be_ignored() {
        CSVHeader header = new CSVHeader(new String[]{"first", "second", null});
        assertThat(header.getColumnNumbers(), is(ImmutableSet.of(0, 1)));
    }

    @Test
    public void duplicate_header_columns_should_be_ignored() {
        CSVHeader header = new CSVHeader(new String[]{"first", "second", "first"});
        assertThat(header.getColumnNumbers(), is(ImmutableSet.of(0, 1)));
    }

    @Test
    public void getColumnName_should_return_header_column() {
        String secondHeaderColumn = "second";
        CSVHeader header = new CSVHeader(new String[]{"first", secondHeaderColumn, "third"});
        assertThat(header.getColumnName(1), is(secondHeaderColumn));
    }

    @Test(expected = IllegalArgumentException.class)
    public void getColumnName_should_throw_Exception_if_no_column_present() {
        CSVHeader header = new CSVHeader(new String[]{"first", "second", "third"});
        header.getColumnName(4);
    }

    @Test
    public void getPathColumn_should_return_header_column_named_path() {
        CSVHeader header = new CSVHeader(new String[]{"first", "path", "third"});

        assertThat(header.getPathColumn(), is(1));
    }

    @Test
    public void getPathColumn_should_return_first_column_if_no_path_colum_present_and_first_column_nonempty() {
        CSVHeader header = new CSVHeader(new String[]{"first", "second", "third"});

        assertThat(header.getPathColumn(), is(0));
    }

    @Test
    public void getPathColumn_should_return_first_nonempty_column_if_no_path_colum_present() {
        CSVHeader header = new CSVHeader(new String[]{"", "second", "third"});

        assertThat(header.getPathColumn(), is(1));
    }
}

File: analysis/import/CSVImporter/src/test/kotlin/de/maibornwolff/codecharta/importer/csv/CSVHeaderTest.kt
package de.maibornwolff.codecharta.importer.csv

import org.hamcrest.CoreMatchers.* // ktlint-disable no-wildcard-imports
import org.hamcrest.MatcherAssert.assertThat
import org.spekframework.spek2.Spek
import org.spekframework.spek2.style.specification.describe
import kotlin.test.assertFailsWith

class CSVHeaderTest : Spek({
    describe("an empty header") {
        it("should throw exception") {
            assertFailsWith(IllegalArgumentException::class) {
                CSVHeader(arrayOf())
            }
        }
    }

    describe("a valid header") {
        val headerLine = arrayOf("first", "second", "", null)
        val header = CSVHeader(headerLine)

        it("should have non-empty columns") {
            assertThat(header.columnNumbers, hasItems(0, 1))
            assertThat(header.getColumnName(0), `is`(headerLine[0]))
            assertThat(header.getColumnName(1), `is`(headerLine[1]))
        }

        it("should ignore empty columns") {
            assertThat(header.columnNumbers, not(hasItem(2)))
        }

        it("should ignore null columns") {
            assertThat(header.columnNumbers, not(hasItem(3)))
        }

        it("getColumnName should throw exception if column name not present") {
            assertFailsWith(IllegalArgumentException::class) {
                header.getColumnName(4)
            }
        }

        it("getPathColumn should return first column if no path column present and first column non empty") {
            assertThat(header.pathColumn, `is`(0))
        }
    }

    describe("a duplicate header") {
        val headerLine = arrayOf<String?>("first", "first")
        val header = CSVHeader(headerLine)

        it("should be ignored") {
            assertThat(header.columnNumbers, hasItems(0))
            assertThat(header.getColumnName(0), `is`(headerLine[0]))
        }
    }

    describe("a header with column with name path") {
        val header = CSVHeader(arrayOf("first", "path", "third"))

        it("should have this column as path column") {
            assertThat(header.pathColumn, `is`(1))
        }
    }

    describe("a header without path column and empty first column") {
        val header = CSVHeader(arrayOf("", null, "second", "third"))

        it("should have first non-empty column as path column") {
            assertThat(header.pathColumn, `is`(2))
        }
    }
})